File: OracleTest/src/DependenciesCollection.java
import java.util.ArrayList;
import java.util.HashMap;

public class DependenciesCollection {
	//public Set<(String,String)> dependencies;
	//description of layers: numberofsentence((wordID, [wordIDs]))
	public ArrayList<HashMap<Integer, ArrayList<Integer>>> leftChildren;
	public ArrayList<HashMap<Integer, ArrayList<Integer>>> rightChildren;
	public ArrayList<HashMap<Integer, Integer>> parents;
	public int numSentences;
	//public HashSet<String> children;
	
	public DependenciesCollection() {
		leftChildren = new ArrayList<HashMap<Integer, ArrayList<Integer>>>();
		rightChildren = new ArrayList<HashMap<Integer, ArrayList<Integer>>>();
		parents = new ArrayList<HashMap<Integer, Integer>>();
		
		leftChildren.add(new HashMap<Integer, ArrayList<Integer>>());
		rightChildren.add(new HashMap<Integer, ArrayList<Integer>>());
		parents.add(new HashMap<Integer, Integer>());
		numSentences = 0;
	}
	
	public HashMap<Integer, ArrayList<Integer>> getLeftChildren(int sentence) {
		return leftChildren.get(sentence);
	}
	
	public HashMap<Integer, ArrayList<Integer>> getRightChildren(int sentence) {
		return rightChildren.get(sentence);
	}
	
	public HashMap<Integer, Integer> getParent(int sentence) {
		return parents.get(sentence);
	}
	
	public void newSentence() {
		leftChildren.add(new HashMap<Integer, ArrayList<Integer>>());
		rightChildren.add(new HashMap<Integer, ArrayList<Integer>>());
		parents.add(new HashMap<Integer, Integer>());
		numSentences++;
	}
	
	public void addAll(ArrayList<ArrayList<String[]>> conllSentences) {
		for (ArrayList<String[]> conllSentence: conllSentences) {
			for (String[] word: conllSentence) {
				Integer parentID = Integer.parseInt(word[CoNLLFieldType.HEAD.ordinal()]);
				//skip the ROOT
				if (parentID >= 0) {
					Integer childID = Integer.parseInt(word[CoNLLFieldType.ID.ordinal()]);
					add(parentID, childID);
				}
			}
		}
	}
	
	public void add(Integer parentID, Integer childID) {
		//dependencies.add((parent[CoNLLFieldType.ID], child[CoNLLFieldType.ID]));
		//System.out.println(parentID);
		//System.out.println(childID);
		parents.get(numSentences).put(childID, parentID);
			
		if (parentID < childID) {
			if (!rightChildren.get(numSentences).containsKey(parentID)) {
				rightChildren.get(numSentences).put(parentID, new ArrayList<Integer>());
			}
			rightChildren.get(numSentences).get(parentID).add(childID);
		}
		else if (parentID > childID) {
			if (!leftChildren.get(numSentences).containsKey(parentID)) {
				leftChildren.get(numSentences).put(parentID, new ArrayList<Integer>());
			}
			leftChildren.get(numSentences).get(parentID).add(childID);
		}
		else {
			//Shouldn't be here
			//Means that parentID == childID, or we're comparing the same words in a sentence
		}
	}
	
	
}

File: zpar/src/common/depparser/dependenciesCollection.cpp
#include "dependenciesCollection.h"

/*DependenciesCollection::DependenciesCollection() {
  leftChildren.push_back(siblings_t());
  rightChildren.push_back(siblings_t());
  parents.push_back(parent_t());
  numSentences = 0;
}*/

siblings_t DependenciesCollection::getLeftChildren(int sentence) {
  return leftChildren.at(sentence);
}

siblings_t DependenciesCollection::getRightChildren(int sentence) {
  return rightChildren.at(sentence);
}

parent_t DependenciesCollection::getParent(int sentence) {
  return parents.at(sentence);
}

void DependenciesCollection::addAll(sentence_t conllSentence) {
  for (unsigned int i=0; i < conllSentence.size(); i++) {
    int parentID = atoi(conllSentence.at(i).at(CONLLFIELD_HEAD).c_str());
    int childID = atoi(conllSentence.at(i).at(CONLLFIELD_ID).c_str());

    add(parentID, childID);
  }
}

void DependenciesCollection::add(int parentID, int childID) {
  parents.at(numSentences).insert(parent_t::value_type(childID, parentID));

  if (childID > parentID) {
	siblings_t rightSiblings = rightChildren.at(numSentences);
	siblings_t::iterator parent_right_it = rightSiblings.find(parentID);
    if (parent_right_it == rightSiblings.end()) {
      rightChildren.at(numSentences).insert(siblings_t::value_type(parentID, std::vector<int>()));
    }
    rightChildren.at(numSentences)[parentID].push_back(childID);
  }
  else if (childID < parentID) {
	siblings_t leftSiblings = leftChildren.at(numSentences);
	siblings_t::iterator parent_left_it = leftSiblings.find(parentID);
    if (parent_left_it == leftSiblings.end()) {
      leftChildren.at(numSentences).insert(siblings_t::value_type(parentID, std::vector<int>()));
    }
    leftChildren.at(numSentences)[parentID].push_back(childID);
  }
  else {
    //Shouldn't be here
	//Means that parentID == childID, or we're comparing the same words in a sentence
  }
}

void DependenciesCollection::newSentence() {
  leftChildren.push_back(siblings_t());
  rightChildren.push_back(siblings_t());
  parents.push_back(parent_t());
  numSentences++;
}
